builtin/providers/rancher/resource_rancher_stack.go

```diff
@@ -148,24 +151,101 @@ func resourceRancherStackUpdate(d *schema.ResourceData, meta interface{}) error
 	if err != nil {
 		return err
 	}
+	d.Partial(true)
 
 	data, err := makeStackData(d, meta)
 	if err != nil {
 		return err
 	}
 
-	// TODO: upgrade stack if docker_compose, rancher_compose or environment have changed
-
-	var newStack rancherClient.Environment
 	stack, err := client.Environment.ById(d.Id())
 	if err != nil {
 		return err
 	}
 
-	if err := client.Update("environment", &stack.Resource, data, &newStack); err != nil {
-		return err
+	if d.HasChange("name") ||
+		d.HasChange("description") {
+
+		stack, err = client.Environment.Update(stack, rancherClient.Environment{
+			Name:        *data["name"].(*string),
+			Description: *data["description"].(*string),
+		})
+		if err != nil {
+			return err
+		}
+
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"active", "upgrading", "upgraded"},
+			Target:     []string{"upgraded"},
+			Refresh:    StackStateRefreshFunc(client, stack.Id),
+			Timeout:    10 * time.Minute,
+			Delay:      1 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+		s, waitErr := stateConf.WaitForState()
+		stack = s.(*rancherClient.Environment)
+		if waitErr != nil {
+			return fmt.Errorf(
+				"Error waiting for stack (%s) to be updated: %s", stack.Id, waitErr)
+		}
+
+		err = resourceRancherStackRead(d, meta)
+		if err != nil {
+			return err
+		}
+
+		d.SetPartial("name")
+		d.SetPartial("description")
+	}
+
+	if d.HasChange("docker_compose") ||
+		d.HasChange("rancher_compose") ||
+		d.HasChange("environment") ||
+		d.HasChange("catalog_id") ||
+		d.HasChange("scope") {
+
+		envMap := make(map[string]interface{})
+		for key, value := range *data["environment"].(*map[string]string) {
+			envMap[key] = &value
+		}
+		stack, err = client.Environment.ActionUpgrade(stack, &rancherClient.EnvironmentUpgrade{
+			DockerCompose:  *data["dockerCompose"].(*string),
+			RancherCompose: *data["rancherCompose"].(*string),
+			Environment:    envMap,
+			ExternalId:     *data["externalId"].(*string),
+		})
+		if err != nil {
+			return err
+		}
+
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"active", "upgrading"},
+			Target:     []string{"active"},
+			Refresh:    StackStateRefreshFunc(client, stack.Id),
+			Timeout:    10 * time.Minute,
+			Delay:      1 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+		s, waitErr := stateConf.WaitForState()
```

Comment:
Should Terraform finish the upgrade when there was no error, and rollback it otherwise?